subsys/net/lib/tls_credentials/tls_credentials_shell.c

```diff
@@ -0,0 +1,805 @@
+
+	{"PRE_SHARED_KEY_ID",	TLS_CREDENTIAL_PSK_ID},
+	{"PSK_ID",		TLS_CREDENTIAL_PSK_ID}
+};
+
+#define ANY_KEYWORD "any"
+
+/* This is so that we can output base64 in chunks of this length if necessary */
+BUILD_ASSERT(
+	(CONFIG_TLS_CREDENTIALS_SHELL_CRED_OUTPUT_WIDTH % 4) == 0,
+	"CONFIG_TLS_CREDENTIALS_SHELL_CRED_OUTPUT_WIDTH must be a multiple of 4."
+);
+
+/* Output buffers used for printing credentials and digests.
+ * One extra byte included for NULL termination.
+ */
+static char cred_out_buf[CONFIG_TLS_CREDENTIALS_SHELL_CRED_OUTPUT_WIDTH + 1];
+static char cred_digest_buf[CONFIG_TLS_CREDENTIALS_SHELL_DIGEST_BUF_SIZE + 1];
+
+/* Internal buffer used for storing and retrieving credentials.
+ * +1 byte for potential NULL termination.
+ */
+static char cred_buf[CONFIG_TLS_CREDENTIALS_SHELL_CRED_BUF_SIZE + 1];
+static size_t cred_written;
+
+/* Some backends (namely, the volatile backend) store a reference rather than a copy of passed-in
+ * credentials. For these backends, we need to copy incoming credentials onto the heap before
+ * attempting to store them.
+ *
+ * Since the backend in use is determind at build time by KConfig, so is this behavior.
+ * If multi/dynamic-backend support is ever added, this will need to be updated.
+ */
+#define COPY_CREDENTIALS_TO_HEAP CONFIG_TLS_CREDENTIALS_BACKEND_VOLATILE
+
+/* Used to track credentials that have been copied permanently to the heap, in case they are
+ * ever deleted and need to be freed.
+ */
+static void *cred_refs[CONFIG_TLS_MAX_CREDENTIALS_NUMBER];
+
+/* Find an empty slot in the cred_refs array, or return -1 if none exists.
+ * Pass NULL to find an unused slot.
+ */
+static int find_ref_slot(void *cred)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(cred_refs); i++) {
+		if (cred_refs[i] == cred) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+/* Helpers */
+
+/* Filter out non-printable characters from a passed-in string of known length */
+static int filter_nonprint(char *buf, size_t len, char inval)
+{
+	int i;
+	int ret = 0;
+
+	for (i = 0; i < len; i++) {
+		if (!isprint((int)buf[i])) {
+			buf[i] = inval;
+			ret = -EINVAL;
+		}
+	}
+
+	return ret;
+}
+
+/* Verify that a provided string consists only of the characters 0-9*/
+static bool check_numeric(char *str)
+{
+	int i;
+	int len = strlen(str);
+
+	for (i = 0; i < len; i++) {
+		if (!isdigit((int)str[i])) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+/* Clear the credential write buffer, returns true if anything was actually cleared. */
+static bool cred_buf_clear(void)
+{
+	bool cleared = cred_written != 0;
+
+	(void)memset(cred_buf, 0, sizeof(cred_buf));
+	cred_written = 0;
+
+	return cleared;
+}
+
+/* Parse a (possibly incomplete) chunk into the credential buffer */
+static int cred_buf_write(char *chunk)
+{
+	char *writehead = cred_buf + cred_written;
+	size_t chunk_len = strlen(chunk);
+
+	/* Verify that there is room for the incoming chunk */
+	if (writehead + chunk_len >= cred_buf + sizeof(cred_buf) - 1) {
```

Comment:
MISRA rule 12.1 would be helpful here: "The precedence of operators within expressions should be made explicit".

Reply:
Sure -- Zephyr seems to tend to not use parentheses in this specific situation (LHS and RHS of a numeric comparator are compound operations consisting exclusively of transitive operations), so I followed suit. But I'll add them if you think it would be helpful

Reply:
Please add them